qiskit/circuit/__init__.py

```diff
@@ -54,81 +78,825 @@
+    circuit is the job of :ref:`Qiskit's transpilation package <qiskit-transpiler>`.
+
+
+.. _circuit-module-api:
+
+API overview of :mod:`qiskit.circuit`
+=====================================
+
+All objects here are described in more detail, and in their greater context in the following
+sections.  This section provides an overview of the API elements documented here.
+
+..
+    TODO: actually write the "in-depth section on building circuits and cross-ref to it.
+
+The principal class is :class:`.QuantumCircuit`, which has its own documentation page, including
+an in-depth section on building circuits. The simplest quantum and classical data is built out of
+"bits" and "registers":
+
+* :class:`Bit`, an atom of data
+    * :class:`Qubit`
+    * :class:`Clbit`
+    * :class:`AncillaQubit`
+
+* :class:`Register`, a collection of bits
+    * :class:`QuantumRegister`
+    * :class:`ClassicalRegister`
+    * :class:`AncillaRegister`
+
+Within a circuit, each complete :class:`CircuitInstruction` is made up of an :class:`Operation`
+(which might be an :class:`Instruction`, a :class:`Gate`, or some other subclass) and the qubit
+and clbit operands.  The core base classes here are:
+
+* :class:`CircuitInstruction`, an operation and its operands
+* :class:`InstructionSet`, a temporary handle to a slice of circuit data
+* :class:`Operation`, any abstract mathematical object or hardware instruction
+    * :class:`AnnotatedOperation`, a subclass with applied abstract modifiers
+        * :class:`InverseModifier`
+        * :class:`ControlModifier`
+        * :class:`PowerModifier`
+
+The most common concrete subclass of the minimal, abstract :class:`Operation` interface is the
+:class:`Instruction`.  While :class:`Operation` can include abstract mathematical objects,
+an :class:`Instruction` is something that could conceivably run directly on hardware. This is in
+turn subclassed by :class:`Gate` and :class:`ControlledGate` that further add unitarity and
+controlled semantics on top:
+
+* :class:`Instruction`, representing a hardware-based instruction
+* :class:`Gate`, representing a hardware instruction that is unitary
+* :class:`ControlledGate`, representing a gate with control structure.
+
+Qiskit includes a large library of standard gates and circuits, which is documented in
+:mod:`qiskit.circuit.library`.  Many of these are declared as Python-object singletons.  The
+machinery for this is described in detail in :mod:`qiskit.circuit.singleton`, of which the main
+classes are each a singleton form of the standard instruction–gate hierarchy:
+
+* :class:`~singleton.SingletonInstruction`
+* :class:`~singleton.SingletonGate`
+* :class:`~singleton.SingletonControlledGate`
+
+Some instructions are particularly special in that they affect the control flow or data flow of the
+circuit.  The top-level ones are:
+
+* :class:`Barrier`, to forbid optimizations from crossing a barrier
```

Comment:
```suggestion
* :class:`Barrier`, defines parts of a the circuit that should be optimized by the transpiler separately, i.e. optimizations spanning over this barrier are not allowed
```

Reply:
Done (mostly) in 2c1cc17232a - I kept it to a single clause up here since we define it more completely in the actual API documentation around it.